### PR TITLE
provider/aws: Allow to change the DynamoDB Endpoint

### LINKED ISSUE #2763: DynamoDB local?
Is there a way to set the [DynamoDB Endpoint](https://github.com/aws/aws-sdk-go/issues/318#issuecomment-122142151) to run table migrations against [dynamodb-local](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.DynamoDBLocal.html)?

### PR DESCRIPTION
**Pull request by phstc**
Hi

This PR allows us to change the DynamoDB Endpoint, which super useful to run "migrations" against [dynamodb-local](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.DynamoDBLocal.html).

With that, we can use the same scripts we use in production (AWS DynamoDB), in development (dynamodb-local).

The way I'm using it:

```bash
cd tf/local

ls
table1.tf
table2.tf
config.tf

terraform plan
terraform apply
```

In the folder `local` besides the table definitions, I have also a `config.tf` with:

```
provider "aws" {
    access_key = "123"
    secret_key = "123"
    region = "us-east-1"
    dynamodb_endpoint = "http://localhost:8000"
}
```

which changes the DynamoDB Endpoint. 

I've intentionally set `access` and `secret` to 123, as we don't need those to connect to dynamodb-local.

Any code feedback would be more than appreciated :beers: 

Fix #2763.

### CODE REVIEW

File: builtin/providers/aws/config.go [hunk 1]
```diff
@@ -84,9 +86,15 @@ func (c *Config) Client() (interface{}, error) {
 			Region:      c.Region,
 			MaxRetries:  c.MaxRetries,
 		}
+		awsDynamoDBConfig := &aws.Config{
+			Credentials: creds,
+			Region:      c.Region,
+			MaxRetries:  c.MaxRetries,
+			Endpoint:    c.DynamoDBEndpoint,
+		}
```
Comment (phstc) [inline builtin/providers/aws/config.go]: I didn't like this duplication of configs, but we can't set `Endpoint` for all other connections. I've also tried to copy `awsConfig`, but go didn't like it:

```go
awsDynamoDBConfig := * awsConfig
awsDynamoDBConfig.Endpoint = c.DynamoDBEndpoint
```

Any ideas?

Comment (phinze) [inline builtin/providers/aws/config.go]: I think this is fine for now - the duplication is super local. If we get more duplication here we can investigate cleanup options. :ok_hand:

File: builtin/providers/aws/provider.go [hunk 2]
```diff
@@ -145,6 +145,13 @@ func Provider() terraform.ResourceProvider {
 					return hashcode.String(v.(string))
 				},
 			},
+
+			"dynamodb_endpoint": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
```
Comment (phinze) [inline builtin/providers/aws/provider.go]: Is `""` equivalent to "use upstream default" here?

**Comment by phinze** [comment]
Thanks for this, @phstc - one inline question, and think you could add the attribute to the [provider docs](https://github.com/hashicorp/terraform/blob/master/website/source/docs/providers/aws/index.html.markdown) too?

File: builtin/providers/aws/provider.go (hunk 2 above)
Comment (phstc) [inline builtin/providers/aws/provider.go]: @phinze that's correct based on their docs: https://github.com/aws/aws-sdk-go/blob/67dc9f948602be9f85cb640f89b0adec994ccbda/aws/config.go#L49-L55 and I've also confirmed that locally.

But I can remove the `Default: ""` above and make it more explicit:

```go
awsDynamoDBConfig := &aws.Config{
	Credentials: creds,
	Region:      c.Region,
	MaxRetries:  c.MaxRetries,
}

if c.DynamoDBEndpoint != nil {
	awsDynamoDBConfig.Endpoint = c.DynamoDBEndpoint
}
```

File: website/source/docs/providers/aws/index.html.markdown [hunk 3]
```diff
@@ -55,5 +55,7 @@ The following arguments are supported in the `provider` block:
   to prevent you mistakenly using a wrong one (and end up destroying live environment).
   Conflicts with `allowed_account_ids`.
 
+* `dynamodb_endpoint` - (Optional) Use this to override the default endpoint URL constructed from the :region. It's typically used to connect to dynamodb-local"
+
```
Comment (phstc) [inline website/source/docs/providers/aws/index.html.markdown]: @phinze provider docs ^^^ updated :beers:

**Comment by phstc** [comment]
@phinze not sure why the tests are failing `make test` is passing locally and the changes I did after the last successfully build were only on the docs.

**Comment by catsby** [comment]
@phstc there have been upstream changes to the `config` struct used by AWS. A simple merge from `master` and changing your config like so should do the trick:

```go
awsDynamoDBConfig := &aws.Config{
  Credentials: creds,
  Region:      aws.String(c.Region),
  MaxRetries:  aws.Int(c.MaxRetries),
  Endpoint:    aws.String(c.DynamoDBEndpoint),
}
```

**Comment by phstc** [comment]
@catsby thanks! all set then:

> All checks have passed

**Comment by phstc** [comment]
:shipit: ?

**Comment by catsby** [comment]
Thanks @phstc !

**Comment by phstc** [comment]
:metal: 

@catsby is there a new release on the schedule? or a mailing list to subscribe to get release updates? I will expect a new release to update my work colleagues to start using terraform with dynamodb-local.

**Comment by catsby** [comment]
>  is there a new release on the schedule

I think next week, I'll double check 

> a mailing list to subscribe to get release updates

There is a mailing list here:

- https://groups.google.com/group/terraform-tool

I don't know that we've been very diligent about announcing there. We typically tweet the announcement here https://twitter.com/hashicorp , e.g. https://twitter.com/hashicorp/status/623250629247520768 

That said, we probably should announce on the list...

**Comment by phstc** [comment]
@catsby thanks! You got a new follower :smile:

**Pull request merged**